File: app/src/main/java/cc/smartcash/wallet/Adapters/TransactionAdapter.java
package cc.smartcash.wallet.Adapters;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

import cc.smartcash.wallet.Models.Coin;
import cc.smartcash.wallet.Models.FullTransaction;
import cc.smartcash.wallet.Models.Transaction;
import cc.smartcash.wallet.R;
import cc.smartcash.wallet.Utils.SmartCashApplication;
import cc.smartcash.wallet.Utils.URLS;
import cc.smartcash.wallet.ViewHolders.TransactionViewHolder;
import cc.smartcash.wallet.ViewModels.TransactionViewModel;

public class TransactionAdapter extends RecyclerView.Adapter<TransactionViewHolder> {


    public static final String TAG = TransactionAdapter.class.getSimpleName();

    private Context context;
    private ArrayList<Transaction> transactions;
    private TransactionViewHolder transactionViewHolder;

    public TransactionAdapter(Context context, ArrayList<Transaction> transactions) {
        this.context = context;
        this.transactions = transactions;
    }

    public void setItems(ArrayList<Transaction> transactions) {
        this.transactions = transactions;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public TransactionViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater.from(context).inflate(R.layout.transaction_item, viewGroup, false);
        transactionViewHolder = new TransactionViewHolder(view);
        return transactionViewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull TransactionViewHolder transactionViewHolder, int i) {

        SmartCashApplication app = new SmartCashApplication(context);
        Coin actualSelectedCoin = app.getActualSelectedCoin(context);

        String fiatValue = "";

        if (actualSelectedCoin == null || actualSelectedCoin.getName().equals(context.getString(R.string.default_crypto))) {
            ArrayList<Coin> currentPrice = app.getCurrentPrice(context);
            fiatValue = (app.formatNumberBySelectedCurrencyCode(app.getCurrentValueByRate(this.transactions.get(i).getAmount(), currentPrice.get(0).getValue())));
        } else {
            fiatValue = (app.formatNumberBySelectedCurrencyCode(app.getCurrentValueByRate(this.transactions.get(i).getAmount(), actualSelectedCoin.getValue())));
        }

        transactionViewHolder.amount.setText(app.formatNumberByDefaultCrypto(this.transactions.get(i).getAmount()));
        transactionViewHolder.direction.setText(this.transactions.get(i).getDirection());
        transactionViewHolder.timestamp.setText(this.transactions.get(i).getTimestamp());
        transactionViewHolder.hash.setText(this.transactions.get(i).getHash());
        transactionViewHolder.price.setText(" (" + fiatValue + ")");
        transactionViewHolder.hash.setOnClickListener(v -> {
            Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(URLS.URL_INSIGHT_EXPLORER + this.transactions.get(i).getHash()));
            context.startActivity(browserIntent);
        });

        transactionViewHolder.btnDetails.setOnClickListener(v -> {
            setVisibility(transactionViewHolder);

            TransactionParameter transactionParameter = new TransactionParameter(transactionViewHolder, transactions.get(i).getHash(), i);
            new TransactionTask().execute(transactionParameter);

        });

        setDirectionColors(transactionViewHolder, i);
    }

    private class TransactionParameter {
        private TransactionViewHolder transactionViewHolder;
        private String hash;
        private int position;
        private FullTransaction transaction;

        public TransactionParameter(TransactionViewHolder transactionViewHolder, String hash, int position) {
            this.transactionViewHolder = transactionViewHolder;
            this.hash = hash;
            this.position = position;
        }

        public FullTransaction getTransaction() {
            return transaction;
        }

        public void setTransaction(FullTransaction transaction) {
            this.transaction = transaction;
        }

        public TransactionViewHolder getTransactionViewHolder() {
            return transactionViewHolder;
        }

        public String getHash() {
            return hash;
        }

        public int getPosition() {
            return position;
        }
    }

    private class TransactionTask extends AsyncTask<TransactionParameter, Integer, TransactionParameter> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override
        protected TransactionParameter doInBackground(TransactionParameter... parameters) {
            parameters[0].getTransactionViewHolder().confirmations.setText("loading...");
            try {
                parameters[0].setTransaction(TransactionViewModel.getSyncTransaction(parameters[0].getHash()));
                return parameters[0];
            } catch (Exception ex) {
                parameters[0].getTransactionViewHolder().confirmations.setText("0");
            }
            return null;
        }

        @Override
        protected void onProgressUpdate(Integer... progress) {
            super.onProgressUpdate(progress);
        }

        @Override
        protected void onPostExecute(TransactionParameter transactionResult) {
            super.onPostExecute(transactionResult);
            try {
                transactionResult.getTransactionViewHolder().confirmations.setText(String.valueOf(transactionResult.getTransaction().getConfirmations()));
            } catch (Exception ex) {
                transactionResult.getTransactionViewHolder().confirmations.setText("0");
            }
        }
    }

    public void setDirectionColors(TransactionViewHolder transactionViewHolder, int i) {
        String direction = String.valueOf(this.transactions.get(i).getDirection());

        if (direction.equals("Sent")) {
            transactionViewHolder.direction.setBackgroundResource(R.drawable.bg_paid);
            Drawable wrappedDrawable = changeIconColor(R.drawable.ic_arrow_up, R.color.paidColor);
            transactionViewHolder.icon.setBackground(wrappedDrawable);
        } else if (direction.equals("Received")) {
            transactionViewHolder.direction.setBackgroundResource(R.drawable.bg_receive);
            Drawable wrappedDrawable = changeIconColor(R.drawable.ic_arrow_down, R.color.receiveColor);
            transactionViewHolder.icon.setBackground(wrappedDrawable);
        } else if (direction.equals("Awaiting")) {
            Drawable wrappedDrawable = changeIconColor(R.drawable.bg_awaiting, R.color.awaitingColor);
            transactionViewHolder.direction.setBackground(wrappedDrawable);

        }
    }

    public void setVisibility(TransactionViewHolder transactionViewHolder) {
        if (transactionViewHolder.idView.getVisibility() == View.VISIBLE) {
            transactionViewHolder.idView.setVisibility(View.GONE);
            transactionViewHolder.btnDetails.setImageResource(R.drawable.menu_down);
        } else {
            transactionViewHolder.idView.setVisibility(View.VISIBLE);
            transactionViewHolder.btnDetails.setImageResource(R.drawable.menu_up);
        }
    }

    public Drawable changeIconColor(Integer icon, Integer color) {
        Drawable wrappedDrawable = DrawableCompat.wrap(ContextCompat.getDrawable(context, icon));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            DrawableCompat.setTint(wrappedDrawable, context.getColor(color));
        }
        return wrappedDrawable;
    }

    @Override
    public int getItemCount() {
        return transactions.size();
    }
}

File: app/src/main/java/cc/smartcash/wallet/Adapters/TransactionAdapter.kt
package cc.smartcash.wallet.Adapters

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.AsyncTask
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.recyclerview.widget.RecyclerView
import cc.smartcash.wallet.Models.FullTransaction
import cc.smartcash.wallet.Models.Transaction
import cc.smartcash.wallet.R
import cc.smartcash.wallet.Utils.SmartCashApplication
import cc.smartcash.wallet.Utils.URLS
import cc.smartcash.wallet.ViewHolders.TransactionViewHolder
import cc.smartcash.wallet.ViewModels.TransactionViewModel
import java.util.*

class TransactionAdapter(private val context: Context, private var transactions: ArrayList<Transaction>?) : RecyclerView.Adapter<TransactionViewHolder>() {
    private var transactionViewHolder: TransactionViewHolder? = null

    fun setItems(transactions: ArrayList<Transaction>) {
        this.transactions = transactions
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): TransactionViewHolder {
        val view = LayoutInflater.from(context).inflate(R.layout.transaction_item, viewGroup, false)
        transactionViewHolder = TransactionViewHolder(view)
        return transactionViewHolder!!
    }

    override fun onBindViewHolder(transactionViewHolder: TransactionViewHolder, i: Int) {

        val app = SmartCashApplication(context)
        val actualSelectedCoin = app.getActualSelectedCoin(context)

        var fiatValue = ""

        fiatValue = if (actualSelectedCoin == null || actualSelectedCoin.name == context.getString(R.string.default_crypto)) {
            val currentPrice = app.getCurrentPrice()
            app.formatNumberBySelectedCurrencyCode(app.getCurrentValueByRate(this.transactions!![i].amount!!, currentPrice!![0].value!!))
        } else {
            app.formatNumberBySelectedCurrencyCode(app.getCurrentValueByRate(this.transactions!![i].amount!!, actualSelectedCoin.value!!))
        }

        transactionViewHolder.amount.text = app.formatNumberByDefaultCrypto(this.transactions!![i].amount!!)
        transactionViewHolder.direction.text = this.transactions!![i].direction
        transactionViewHolder.timestamp.text = this.transactions!![i].timestamp
        transactionViewHolder.hash.text = this.transactions!![i].hash
        transactionViewHolder.price.text = " ($fiatValue)"
        transactionViewHolder.hash.setOnClickListener { v ->
            val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(URLS.URL_INSIGHT_EXPLORER + this.transactions!![i].hash!!))
            context.startActivity(browserIntent)
        }

        transactionViewHolder.btnDetails.setOnClickListener { v ->
            setVisibility(transactionViewHolder)

            val transactionParameter = TransactionParameter(transactionViewHolder, transactions!![i].hash!!, i)
            TransactionTask().execute(transactionParameter)

        }

        setDirectionColors(transactionViewHolder, i)
    }

    private inner class TransactionParameter(val transactionViewHolder: TransactionViewHolder, val hash: String, val position: Int) {
        var transaction: FullTransaction? = null
    }

    private inner class TransactionTask : AsyncTask<TransactionParameter, Int, TransactionParameter>() {

        override fun doInBackground(vararg parameters: TransactionParameter): TransactionParameter? {
            (context as Activity).runOnUiThread { parameters[0].transactionViewHolder.confirmations.text = "loading..." }
            try {
                val syncTransaction = TransactionViewModel.getSyncTransaction(parameters[0].hash)
                parameters[0].transaction = syncTransaction
                return parameters[0]
            } catch (ex: Exception) {
                context.runOnUiThread { parameters[0].transactionViewHolder.confirmations.text = "0" }
            }

            return null
        }

        override fun onPostExecute(transactionResult: TransactionParameter) {
            super.onPostExecute(transactionResult)
            try {
                transactionResult.transactionViewHolder.confirmations.text = transactionResult.transaction!!.confirmations.toString()
            } catch (ex: Exception) {
                transactionResult.transactionViewHolder.confirmations.text = "0"
            }

        }
    }

    private fun setDirectionColors(transactionViewHolder: TransactionViewHolder, i: Int) {

        when (transactions!![i].direction.toString()) {
            "Sent" -> {
                transactionViewHolder.direction.setBackgroundResource(R.drawable.bg_paid)
                val wrappedDrawable = changeIconColor(R.drawable.ic_arrow_up, R.color.paidColor)
                transactionViewHolder.icon.background = wrappedDrawable
            }
            "Received" -> {
                transactionViewHolder.direction.setBackgroundResource(R.drawable.bg_receive)
                val wrappedDrawable = changeIconColor(R.drawable.ic_arrow_down, R.color.receiveColor)
                transactionViewHolder.icon.background = wrappedDrawable
            }
            "Awaiting" -> {
                val wrappedDrawable = changeIconColor(R.drawable.bg_awaiting, R.color.awaitingColor)
                transactionViewHolder.direction.background = wrappedDrawable

            }
        }
    }

    private fun setVisibility(transactionViewHolder: TransactionViewHolder) {
        if (transactionViewHolder.idView.visibility == View.VISIBLE) {
            transactionViewHolder.idView.visibility = View.GONE
            transactionViewHolder.btnDetails.setImageResource(R.drawable.menu_down)
        } else {
            transactionViewHolder.idView.visibility = View.VISIBLE
            transactionViewHolder.btnDetails.setImageResource(R.drawable.menu_up)
        }
    }

    private fun changeIconColor(icon: Int?, color: Int?): Drawable {
        val wrappedDrawable = DrawableCompat.wrap(ContextCompat.getDrawable(context, icon!!)!!)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            DrawableCompat.setTint(wrappedDrawable, context.getColor(color!!))
        }
        return wrappedDrawable
    }

    override fun getItemCount(): Int {
        return transactions!!.size
    }

    companion object {


        val TAG: String? = TransactionAdapter::class.java.simpleName
    }
}
